src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java

```diff
@@ -74,28 +76,36 @@ public class HTTPRecordWriter extends RecordWriter<StructuredRecord, StructuredR
   public void write(StructuredRecord input, StructuredRecord unused) throws IOException {
     String message = null;
     if (config.getMethod().equals("POST") || config.getMethod().equals("PUT")) {
-      if (config.getMessageFormat().equals("JSON")) {
-        message = StructuredRecordStringConverter.toJsonString(input);
-        contentType = "application/json";
-      } else if (config.getMessageFormat().equals("Form")) {
-        message = createFormMessage(input);
-        contentType = " application/x-www-form-urlencoded";
-      } else if (config.getMessageFormat().equals("Custom")) {
-        message = createCustomMessage(config.getBody(), input);
-        contentType = " text/plain";
+      switch (config.getMessageFormat()) {
+        case "JSON":
+          message = StructuredRecordStringConverter.toJsonString(input);
+          contentType = "application/json";
+          break;
+        case "Form":
+          message = createFormMessage(input);
+          contentType = "application/x-www-form-urlencoded";
+          break;
+        case "Custom":
+          message = createCustomMessage(config.getBody(), input);
+          contentType = "text/plain";
+          break;
+        default:
+          throw new IllegalArgumentException("Invalid message format.");
       }
-      messages.append(message).append(config.getDelimiterForMessages());
+      messageBuffer.add(message);
     }
-    StringTokenizer tokens = new StringTokenizer(messages.toString().trim(), config.getDelimiterForMessages());
-    if (config.getBatchSize() == 1 || tokens.countTokens() == config.getBatchSize()) {
+    int batchSize = config.getBatchSize();
+    if (messageBuffer.size() == batchSize) {
+      flushMessageBuffer();
       executeHTTPService();
     }
   }
 
   @Override
   public void close(TaskAttemptContext taskAttemptContext) throws IOException, InterruptedException {
     // Process remaining messages after batch executions.
-    if (!messages.toString().isEmpty()) {
+    if (!messageBuffer.isEmpty()) {
+      flushMessageBuffer();
```

Comment:
this is an existing issue a few lines below, but it's not good to catch the Exception from executeHTTPService and rethrow it as a RuntimeException without setting the cause. This results in a bad exception message and loses the stack trace.

It should just call executeHTTPService() without catching, as this method throws IOException already.